espi: MEC172x: updated eSPI virtual wires to use DT

jvasanth1 opened the pull request:
Modify Mircrochip MEC172x eSPI driver to get eSPI virtual wire hardware routing from device tree.

jvasanth1 committed:
drivers: espi: Update Microchip MEC172x eSPI virtual wires to use DT

Modify Mircrochip MEC172x eSPI driver to get eSPI virtual wire
hardware routing from device tree.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
samples: drivers: espi: update MEC172x espi sample

MEC172x espi sample to use eSPI virtual wire hardware routing
through device tree

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

jvasanth1 committed:
drivers: espi: Microchip MEC172x eSPI VW initialization update

Change device tree VW routing to a form allowing overrides.
Add two new DT optional properties for specifying the reset
source and reset value of each virtual wire. Only virtual
wires that are enabled using the status property are modified.
NOTE: eSPI virtual wires are controlled in groups of 4 by
hardware. The optional reset signal source properties applies
to all four virtual wires in the group. If this field is
changed from the hardware default, it should be changed for
only one virtual wire in the group. If the property exists
in more than one wire in the group it must be set to the
same value.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

albertofloyd commented on drivers/espi/espi_mchp_xec_v2.c:
```diff
@@ -97,74 +97,53 @@ LOG_MODULE_REGISTER(espi, CONFIG_ESPI_LOG_LEVEL);
+ * These are configurable by overriding device tree vw routing             |
+ *  50h   | SMVW06 | gpio_3       | gpio_2       | gpio_1    | gpio_0      |
+ *  51h   | SMVW07 | gpio_7       | gpio_6       | gpio_5    | gpio_4      |
+ *  52h   | SMVW08 | ocb_3        | ocb_2        | ocb_1     | ocb_0       |
```
@jvasanth1 this is index mapping is not correct 
![image](https://user-images.githubusercontent.com/16172078/219719410-ee8054ad-ae57-44dc-85b5-467bab84f8b1.png)

albertofloyd commented on drivers/espi/espi_mchp_xec_v2.c:
```diff
@@ -1169,27 +1162,81 @@ static void vw_pltrst_handler(int girq_id, int src, void *user)
+{
+	const struct device *dev = (const struct device *)user;
+
+	notify_system_state(dev, ESPI_VWIRE_SIGNAL_SUS_STAT);
```
@jvasanth1 none of these VW reflect a system state so these should "notify_host_warning"

jvasanth1 replied:
ok will update to invoke `notify_host_warning` instead; 
do we need to update the function `notify_host_warning` to handle `ESPI_VWIRE_SIGNAL_SUS_STAT`

albertofloyd replied:
If you mean driver automatic handling acknowledge, no it's not required.

jvasanth1 replied:
Thanks,
updated the code

albertofloyd commented on dts/arm/microchip/mec172x/mec172x-vw-routing.dtsi:
```diff
@@ -4,111 +4,228 @@
+			vw-reg = <0x51 SMVW 7 3>;
+			status = "disabled";
+		};
+		vw_t2c_gpio_8: vw_t2c_gpio_8 {
```
@jvasanth1 OCB host index is incorrect as indicated on the documented table. This should be 50h

MartinYan-mchp replied:
per following 3 information:

////
1:
vw_t2c_gpio_8: vw_t2c_gpio_8 {
vw-reg = <0x52 SMVW 8 0>;
status = "disabled";
};
2:
MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_OCB_0, vw_t2c_gpio_8),
3:
*  52h   | SMVW08 | ocb_3        | ocb_2        | ocb_1     | ocb_0       |
////

OCB_0 is located at Index 52h bit0.

do you request to move OCB_0 to Index 50h bit0?

MartinYan-mchp replied:
Porvide one more information from Chrome EC project:

https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/3863940
__override void board_overcurrent_event(int port, int is_overcurrented)
{
	/*
	 * Meteorlake PCH uses Virtual Wire for over current error,
	 * hence Send 'Over Current Virtual Wire' eSPI signal.
	 */
	espi_send_vwire(espi_dev, port + ESPI_VWIRE_SIGNAL_SLV_GPIO_0,
			!is_overcurrented);
}

ESPI_VWIRE_SIGNAL_SLV_GPIO_0 is used for overcurrent, i.e. VW Index 50h bit0.
	ESPI_VWIRE_SIGNAL_OCB_0,
	ESPI_VWIRE_SIGNAL_OCB_1,
	ESPI_VWIRE_SIGNAL_OCB_2,
	ESPI_VWIRE_SIGNAL_OCB_3,
in driver should be removed and redefined in application level?

vphirema replied:
Yes, From the eSPI spec looks like there is no mention about OCB. OCB is more of a platform specific.
Non Intel SOC with eSPI can use it for some other purpose.

albertofloyd replied:
Sounds like ESPI_VWIRE_SIGNAL_SLV_GPIO_0 is same than ESPI_VWIRE_SIGNAL_OCB_0 both are intended to be transmitted over current status and used host index 50h. 

So having OCB in 50h and 52h is not ok.

Option #1: 
ESPI_VWIRE_SIGNAL_OCB_0 existed first as platform-specific virtual wire, so I'm thinking we should keep that.

Option #2: Remove ESPI_VWIRE_SIGNAL_OCB_0 altogether from drivers/espi.h 
This breaks current apps.

Option #3: Alternatively we could probably have some map definitions in drivers/espi.h 
#define ESPI_VWIRE_SIGNAL_OCB_0  ESPI_VWIRE_SIGNAL_SLV_GPIO_0 . 
#define ESPI_VWIRE_SIGNAL_OCB_1  ESPI_VWIRE_SIGNAL_SLV_GPIO_1.

I'm inclined for #1 or #3. But in all options this implies to cleanup this initialization table to avoid the duplicate custom virtual wires.

vphirema replied:
cc: @keith-zephyr 

Is there a plan to make ESPI_VWIRE_SIGNAL_OCB_x as VW name in eSPI spec?
if so, we can remove the ESPI_VWIRE_SIGNAL_SLV_GPIO_x names drivers/espi.h and rename them as ESPI_VWIRE_SIGNAL_OCB_x in NPCX code.
else we can remove ESPI_VWIRE_SIGNAL_OCB_x in drivers/espi.h and rename them as ESPI_VWIRE_SIGNAL_SLV_GPIO_x in MCHP code.

Naming VWs at 51h with ESPI_VWIRE_SIGNAL_SLV_GPIO_3 may look odd.

Option 3 at At the application code looks good

albertofloyd replied:
@vphirema you are missing the main point that is there is duplication of "mappings" in the table for VW that is intended to convey the same info. 

@MartinYan-mchp please let's start by correcting the table. I'm ok if you leave the  ESPI_VWIRE_SIGNAL_SLV_GPIO_x names in said table if this PR includes the change proposed in drivers/espi.h

jvasanth1 replied:
Option 3 looks reasonable

Keep all naming as `ESPI_VWIRE_SIGNAL_SLV_GPIO_x`

and add mapping in drivers/espi.h
```
#define ESPI_VWIRE_SIGNAL_OCB_0 ESPI_VWIRE_SIGNAL_SLV_GPIO_8
#define ESPI_VWIRE_SIGNAL_OCB_1 ESPI_VWIRE_SIGNAL_SLV_GPIO_9
#define ESPI_VWIRE_SIGNAL_OCB_2 ESPI_VWIRE_SIGNAL_SLV_GPIO_10
#define ESPI_VWIRE_SIGNAL_OCB_3 ESPI_VWIRE_SIGNAL_SLV_GPIO_11
```

jvasanth1 replied:
updated

vphirema commented on drivers/espi/espi_mchp_xec_v2.c:
```diff
@@ -97,74 +97,53 @@ LOG_MODULE_REGISTER(espi, CONFIG_ESPI_LOG_LEVEL);
+	MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_SLV_GPIO_5, vw_c2t_gpio_5),
+	MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_SLV_GPIO_6, vw_c2t_gpio_6),
+	MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_SLV_GPIO_7, vw_c2t_gpio_7),
+	MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_OCB_0, vw_c2t_gpio_8),
```
As 50h & 5Fh as platform specific slave to master VW.
How about using generic name "ESPI_VWIRE_SIGNAL_SLV_GPIO_x" from eSPI spec
instead of "ESPI_VWIRE_SIGNAL_OCB_x"  as in 
https://github.com/zephyrproject-rtos/zephyr/pull/49710/

Table 2-14. PCH/SoC eSPI Virtual Wires Mapping (Address 50h to 5Fh)
https://cdrdv2.intel.com/v1/dl/getContent/562633?explicitVersion=true

jvasanth1 replied:
updated to use the generic name in the` espi_vwire_signal` enum in` include/zephyr/drivers/espi.h`

vphirema replied:
with the new change ESPI_VWIRE_SIGNAL_OCB_0 = ESPI_VWIRE_SIGNAL_SLV_GPIO_0
should these be ESPI_VWIRE_SIGNAL_SLV_GPIO_8 and above?

albertofloyd reviewed (changes requested):
Need to close on host index divergence.

albertofloyd commented on drivers/espi/espi_mchp_xec_v2.c:
```diff
@@ -1301,6 +1348,103 @@ static void espi_xec_connect_irqs(const struct device *dev)
+		if (rst_state) { /* change reset state and default value of this vwire? */
+			rst_state--;
+			if (rst_state) {
+				temp |= BIT(p->bit + 4);
```
4u

jvasanth1 replied:
updated

albertofloyd commented on drivers/espi/espi_mchp_xec_v2.c:
```diff
@@ -1301,6 +1348,103 @@ static void espi_xec_connect_irqs(const struct device *dev)
+ */
+static void xec_vw_config(const struct device *dev)
+{
+	for (int i = 0; i < ARRAY_SIZE(vw_tbl); i++) {
```
@MartinYan-mchp @jvasanth1 this should not start zero but only take care of the customizable or platform-specific virtual wires.

The rest of eSPI VW have well defined initial, host index mappings. 
Allowing to change them via device tree is probably in detriment may be not best approach.

jvasanth1 replied:
should we start this config from` ESPI_VWIRE_SIGNAL_SLV_GPIO_0` and keep the remaining (spec-defined) VW hard-coded in the driver (such that they don't come from the DTS)

jvasanth1 replied:
updated to start iterating the for loop from `ESPI_VWIRE_SIGNAL_SLV_GPIO_0` instead of 0

jvasanth1 committed:
include: espi.h: Update OCB virtual wires mappings

update espi_vwire_signal enum to refer all platform specific
usage virtual wires to ESPI_VWIRE_SIGNAL_SLV_GPIO_x
Virtual wires used for USB-C port over current (OCB)
have been defined as macros mapped to the corresponding
espi_vwire_signal enum

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

albertofloyd commented:
@jvasanth1 looks good to me, just needs to take care of the compliance checks.

albertofloyd commented on drivers/espi/espi_mchp_xec_v2.c:
```diff
@@ -97,74 +97,53 @@ LOG_MODULE_REGISTER(espi, CONFIG_ESPI_LOG_LEVEL);
  *  47h   | MSVW07 | res          | res          | res       | HOST_C10    |
  *  4Ah   | MSVW08 | res          | res          | DNX_WARN  | res         |
+ * These are configurable by overriding device tree vw routing             |
+ *  50h   | SMVW06 | gpio_3       | gpio_2       | gpio_1    | gpio_0      |
```
@jvasanth1 minor correction required here
***  50h   | SMVW06 | ocb_3        | ocb_2        | ocb_1     | ocb_0       |**
 *  51h   | SMVW07 | gpio_7       | gpio_6       | gpio_5    | gpio_4      |
 ***  52h   | SMVW08 | gpio_11        | gpio_10 | gpio_9  | gpio_8     |**

jvasanth1 replied:
updated

vphirema commented on drivers/espi/espi_mchp_xec_v2.c:
```diff
@@ -132,18 +132,18 @@ static const struct xec_signal vw_tbl[] = {
+	MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_SLV_GPIO_5, vw_t2c_gpio_5),
+	MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_SLV_GPIO_6, vw_t2c_gpio_6),
+	MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_SLV_GPIO_7, vw_t2c_gpio_7),
+	MCHP_DT_ESPI_VW_ENTRY(ESPI_VWIRE_SIGNAL_OCB_0, vw_t2c_gpio_8),
```
with the new change ESPI_VWIRE_SIGNAL_OCB_0 = ESPI_VWIRE_SIGNAL_SLV_GPIO_0
should these be ESPI_VWIRE_SIGNAL_SLV_GPIO_8 and above?

jvasanth1 replied:
this table is generic, doesn't know of the intended mapping, so has all the vw gpios in the enum `espi_vwire_signal`

vphirema replied:
ACK

vphirema commented on include/zephyr/drivers/espi.h:
```diff
@@ -242,17 +242,21 @@ enum espi_vwire_signal {
-	ESPI_VWIRE_SIGNAL_OCB_1,
-	ESPI_VWIRE_SIGNAL_OCB_2,
-	ESPI_VWIRE_SIGNAL_OCB_3,
+	ESPI_VWIRE_SIGNAL_SLV_GPIO_8,
```
is Windows using ESPI_VWIRE_SIGNAL_SLV_GPIO_8 and above? if not we can remove these

albertofloyd replied:
No, but since generic custom VWs mechanism is part of this PR scope, I see no harm

vphirema replied:
I don't a have strong opinion, but if unused it looks like we are spamming header file

jvasanth1 replied:
I agree with @albertofloyd that it is no harm to keep them; I don't see it as spamming the header file; it may have use later

vphirema replied:
ACK
